domain/src/SharePooling/SharePooling.php

```diff
@@ -0,0 +1,193 @@
+<?php
+
+namespace Domain\SharePooling;
+
+use Domain\Enums\FiatCurrency;
+use Domain\SharePooling\Actions\AcquireSharePoolingToken;
+use Domain\SharePooling\Actions\DisposeOfSharePoolingToken;
+use Domain\SharePooling\Actions\RevertSharePoolingTokenDisposal;
+use Domain\SharePooling\Events\SharePoolingTokenAcquired;
+use Domain\SharePooling\Events\SharePoolingTokenDisposalReverted;
+use Domain\SharePooling\Events\SharePoolingTokenDisposedOf;
+use Domain\SharePooling\Exceptions\SharePoolingException;
+use Domain\SharePooling\Services\DisposalProcessor;
+use Domain\SharePooling\Services\QuantityAdjuster;
+use Domain\SharePooling\Services\ReversionFinder;
+use Domain\SharePooling\ValueObjects\QuantityBreakdown;
+use Domain\SharePooling\ValueObjects\SharePoolingTokenAcquisition;
+use Domain\SharePooling\ValueObjects\SharePoolingTokenDisposal;
+use Domain\SharePooling\ValueObjects\SharePoolingTokenDisposals;
+use Domain\SharePooling\ValueObjects\SharePoolingTransactions;
+use EventSauce\EventSourcing\AggregateRoot;
+use EventSauce\EventSourcing\AggregateRootBehaviour;
+use EventSauce\EventSourcing\AggregateRootId;
+
+final class SharePooling implements AggregateRoot
+{
+    use AggregateRootBehaviour;
+
+    private ?FiatCurrency $fiatCurrency = null;
+    private SharePoolingTransactions $transactions;
+
+    private function __construct(AggregateRootId $aggregateRootId)
+    {
+        $this->aggregateRootId = $aggregateRootId;
+
+        $this->transactions = SharePoolingTransactions::make();
+    }
+
+    /** @throws SharePoolingException */
+    public function acquire(AcquireSharePoolingToken $action): void
+    {
+        if ($this->fiatCurrency && $this->fiatCurrency !== $action->costBasis->currency) {
+            throw SharePoolingException::cannotAcquireFromDifferentFiatCurrency(
+                sharePoolingId: $action->sharePoolingId,
+                from: $this->fiatCurrency,
+                to: $action->costBasis->currency,
+            );
+        }
+
+        $disposalsToRevert = ReversionFinder::disposalsToRevertOnAcquisition(
+            acquisition: $action,
+            transactions: $this->transactions,
+        );
+
+        $this->revertDisposals($action->sharePoolingId, $disposalsToRevert);
+
+        // Record the new acquisition
+        $this->recordThat(new SharePoolingTokenAcquired(
+            sharePoolingId: $action->sharePoolingId,
+            sharePoolingTokenAcquisition: new SharePoolingTokenAcquisition(
+                date: $action->date,
+                quantity: $action->quantity,
+                costBasis: $action->costBasis,
+            ),
+        ));
```

Comment:
ℹ️ We need to record the new acquisition before replaying the reverted disposals, as the latter need to be aware of the former so their cost basis can be calculated appropriately